### PR TITLE
usePrettyWriter and precision options for JSON writing

### PR DESCRIPTION
**Pull request by eundersander**
## Motivation and Context

Turning off the PrettyWriter and reducing precision helps reduce filesize for gfx-replay and `arrange_recorder`.

I also made a small change to how Quaternions are serialized, again to reduce filesize.

## How Has This Been Tested

GfxReplayTest was modified to ensure we test precision of serialized/deserialized Quaternions.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/tests/GfxReplayTest.cpp [hunk 1]
```diff
@@ -357,6 +357,20 @@ TEST(GfxReplayTest, simulatorIntegration) {
   simConfig.enableGfxReplaySave = true;
 
   auto sim = Simulator::create_unique(simConfig);
+  auto objAttrMgr = sim->getObjectAttributesManager();
+  objAttrMgr->loadAllJSONConfigsFromPath(
+      Cr::Utility::Directory::join(TEST_ASSETS, "objects/nested_box"), true);
+
+  auto handles = objAttrMgr->getObjectHandlesBySubstring("nested_box");
```
Comment (eundersander) [inline src/tests/GfxReplayTest.cpp:364]: Related to the JSON changes: I'm modifying this test to add a render object instance and verify its pose gets serialized/deserialized with reasonable precision.

File: src/esp/gfx/replay/Recorder.cpp [hunk 2]
```diff
@@ -162,9 +162,13 @@ void Recorder::advanceKeyframe() {
   currKeyframe_ = Keyframe{};
 }
 
-void Recorder::writeSavedKeyframesToFile(const std::string& filepath) {
+void Recorder::writeSavedKeyframesToFile(const std::string& filepath,
+                                         bool usePrettyWriter) {
   auto document = writeKeyframesToJsonDocument();
-  esp::io::writeJsonToFile(document, filepath);
+  // replay::Keyframes use floats (not doubles) so this is plenty of precision
+  const float maxDecimalPlaces = 7;
```
Comment (jturner65) [inline src/esp/gfx/replay/Recorder.cpp:169]: Maybe make an argument with a default value? (standard magic number stuff, not a big deal)

File: src/esp/io/JsonUtils.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#ifndef ESP_IO_JSONUTILS_H_
+#define ESP_IO_JSONUTILS_H_
+
+namespace esp {
+namespace io {
+
+// when writing to json, we often want to clamp tiny floats to 0.0 so they can
+// be serialized concisely as "0.0"
+template <typename T>
+T squashTinyFloats(T x) {
```
Comment (jturner65) [inline src/esp/io/JsonUtils.h:14]: Since we are supporting doubles with this, maybe "squashTinyDecimals" ?

**Review by jturner65 — APPROVED** [review]
LGTM.  Left a couple of comments on some minor things (you can take or leave), but otherwise looks great.

File: src/esp/gfx/replay/Recorder.cpp (hunk 2 above)
Comment (eundersander) [reply src/esp/gfx/replay/Recorder.cpp:169]: I don't want to let the user make a bad decision here like setting it to 0 or 2. This is a decision for Recorder.

**Pull request merged**